File: test/conformance/java/rows/W7.java
// W7.java — "arg and meta are two lanes" (cell java/W7).
//
// The spec: `arg` is the guard's ordinal and `meta` is the op's own
// argument (a text flavour). They are two facts that must never share
// one lane (docs/FIXED-FORM-ALGORITHM.md:245). A string(N) under a
// union arm needs both: the arm ordinal for the guard, and the flavour
// (1=utf8) for the text read.
//
// UT1.schema puts a string(8) under the SECOND arm (ordinal 2, flavour
// 1). The test writes a UT1 file with arm b's label = "hello", reads it
// through a compiled plan that guards b's fields on ordinal 2 while the
// reader's own tag is ordinal 3, and verifies the label survives as
// UTF-8.
//
// If arg and meta shared one lane, the flavour (1) would overwrite the
// guard ordinal (2) and the text entry would either not fire or read as
// WIDE.

package rows;

import tblut.*;
import tblut.TableFixed;

public final class W7 {
    private W7() {}

    // ---- the UT1 wire vector: one record, arm b, label "hello"
    //
    // Built using the generated tblut writers. The file is form 3 with
    // the UT1 layout hash and one record.

    private static byte[] buildUt1File() {
        byte[] body = new byte[UtRootFixed.bodyBytes];
        UtPickFixed.Value pick = new UtPickFixed.Value();
        pick.type = UtPickFixed.b; // arm b, ordinal 2
        byte[] hello = "hello".getBytes(java.nio.charset.StandardCharsets.UTF_8);
        System.arraycopy(hello, 0, pick.b.label, 0, hello.length);
        pick.b.labelLength = hello.length;
        pick.b.grade = 0; // bronze
        UtPickFixed.writeBody(body, 0, pick);
        TableFixed.put32(body, 18, 3); // tail = 3

        int headerLen = UtRootFixed.headerBytes;
        int totalLen = headerLen + UtRootFixed.recordBytes;
        byte[] file = new byte[totalLen];
        UtRootFixed.writeHeader(file);
        TableFixed.put64(file, headerLen, UtRootFixed.hash);
        System.arraycopy(body, 0, file, headerLen + 8, UtRootFixed.bodyBytes);
        return file;
    }

    // ---- a compiled plan that reads UT1's body
    //
    // Record body layout (after the 8-byte hash):
    //   offset  0: pick tag (1 byte) — value 2 means arm b
    //   offset  1: pick body (17 bytes)
    //     offset  1: b.labelLength (4 bytes, int32 LE)
    //     offset  5: b.label (up to 8 bytes, UTF-8)
    //     offset 13: b.grade (1 byte)
    //     offset 14: b.m (4 bytes, int32 LE)
    //   offset 18: tail (4 bytes, int32 LE)
    //
    // Image layout (same as UtRootFixed):
    //   offset  0: pick tag (1 byte)
    //   offset  1: pick body (17 bytes)
    //     offset  1: b.labelLength (4 bytes)
    //     offset  5: b.label (8 bytes)
    //     offset 13: b.grade (1 byte)
    //     offset 14: b.m (4 bytes)
    //   offset 18: tail (4 bytes)
    //
    // The opText entry:
    //   src = offset of the LENGTH word in the source (5)
    //   dst = offset of the LENGTH word in the image (1)
    //   aux = offset of the PAYLOAD in the image (5)
    //   size = payload byte span (8)

    private static TableFixed.Entry[] buildCompiledPlan() {
        TableFixed.Entry[] plan = TableFixed.plan(5);
        int guardOffset = 0; // tag is at offset 0 in the union body
        long writerBOrdinal = 2; // UT1's ordinal for b

        // Entry 0: copy the tag (1 byte at src=0 to dst=0) — unguarded
        plan[0].src = 0; plan[0].dst = 0; plan[0].size = 1;
        plan[0].op = TableFixed.opCopy;
        plan[0].guard = TableFixed.noGuard;

        // Entry 1: label — text op, guarded on writer's ordinal 2
        // src=1 (length word in writer's record, right after pick tag),
        // dst=1 (length word in image, right after pick tag),
        // size=8 (payload byte span), aux=5 (payload destination in image)
        // meta=textUtf8 (1), the flavour — NOT the arm ordinal
        plan[1].src = 1; plan[1].dst = 1; plan[1].size = 8; plan[1].aux = 5;
        plan[1].op = TableFixed.opText;
        plan[1].guard = guardOffset;
        plan[1].arg = writerBOrdinal;
        plan[1].meta = TableFixed.textUtf8;

        // Entry 2: grade (1 byte, src=13, dst=13) — guarded
        plan[2].src = 13; plan[2].dst = 13; plan[2].size = 1;
        plan[2].op = TableFixed.opCopy;
        plan[2].guard = guardOffset;
        plan[2].arg = writerBOrdinal;

        // Entry 3: m (4 bytes, src=14, dst=14) — guarded
        plan[3].src = 14; plan[3].dst = 14; plan[3].size = 4;
        plan[3].op = TableFixed.opCopy;
        plan[3].guard = guardOffset;
        plan[3].arg = writerBOrdinal;

        // Entry 4: tail (4 bytes, src=18, dst=18) — unguarded
        plan[4].src = 18; plan[4].dst = 18; plan[4].size = 4;
        plan[4].op = TableFixed.opCopy;
        plan[4].guard = TableFixed.noGuard;

        return plan;
    }

    private static String readLabel(UtArmBFixed.Value b) {
        return new String(b.label, 0, b.labelLength,
                java.nio.charset.StandardCharsets.UTF_8);
    }

    public static void main(String[] args) {
        int pass = 0;
        int fail = 0;

        // ---- build the UT1 file
        byte[] file = buildUt1File();

        // ---- identity plan read (should always work)
        byte[] image = UtRootFixed.image();
        UtRootFixed.Value result = new UtRootFixed.Value();
        TableFixed.Report report = new TableFixed.Report();
        short[] remap = new short[256];

        int n = UtRootFixed.load(new UtRootFixed.Value[]{result}, 1, file,
                UtRootFixed.identityPlan(), remap, image, report);
        if (n != 1) {
            System.out.println("FAIL identity load returned " + n);
            fail++;
        } else if (report.malformed || report.refused) {
            System.out.println("FAIL identity load: malformed=" + report.malformed
                    + " refused=" + report.refused);
            fail++;
        } else {
            String label = readLabel(result.pick.b);
            if ("hello".equals(label) && result.tail == 3) {
                System.out.println("PASS identity read: label=hello, tail=3");
                pass++;
            } else {
                System.out.println("FAIL identity read: label=" + label + " tail=" + result.tail);
                fail++;
            }
        }

        // ---- compiled plan read: the critical test
        //
        // The plan guards b's fields on the writer's ordinal (2), and the
        // text entry uses meta=textUtf8 (1). The two lanes must be separate.
        TableFixed.Entry[] plan = buildCompiledPlan();
        byte[] image2 = UtRootFixed.image();
        TableFixed.Report report2 = new TableFixed.Report();

        int headerLen = UtRootFixed.headerBytes;
        byte[] recordBody = new byte[UtRootFixed.bodyBytes];
        System.arraycopy(file, headerLen + 8, recordBody, 0, UtRootFixed.bodyBytes);

        TableFixed.run(plan, 5, remap, recordBody, 0, recordBody.length, image2, report2);

        if (report2.malformed) {
            System.out.println("FAIL compiled plan: malformed");
            fail++;
        } else if (report2.refused) {
            System.out.println("FAIL compiled plan: refused " + report2.reason);
            fail++;
        } else {
            UtRootFixed.Value result2 = new UtRootFixed.Value();
            UtRootFixed.scatter(image2, 0, result2, report2);

            String label2 = readLabel(result2.pick.b);
            if ("hello".equals(label2)) {
                System.out.println("PASS compiled plan read: label=hello (arg and meta are two lanes)");
                pass++;
            } else {
                System.out.println("FAIL compiled plan read: label=" + label2
                        + " (expected hello; arg/meta likely shared a lane)");
                fail++;
            }
        }

        // ---- negative control: break the text flavour
        //
        // Set meta to textWide instead of textUtf8. A wide read halves the
        // bound (4 units instead of 8 bytes) and reads UTF-16 code units,
        // which will not match "hello" encoded as UTF-8 bytes.
        TableFixed.Entry[] badPlan = buildCompiledPlan();
        badPlan[1].meta = TableFixed.textWide;
        byte[] image3 = UtRootFixed.image();
        TableFixed.Report report3 = new TableFixed.Report();

        TableFixed.run(badPlan, 5, remap, recordBody, 0, recordBody.length, image3, report3);
        UtRootFixed.Value result3 = new UtRootFixed.Value();
        UtRootFixed.scatter(image3, 0, result3, report3);

        String label3 = readLabel(result3.pick.b);
        if (!"hello".equals(label3)) {
            System.out.println("PASS negative control: wrong flavour misreads label (" + label3 + ")");
            pass++;
        } else {
            System.out.println("FAIL negative control: wrong flavour still reads correctly");
            fail++;
        }

        System.out.println(pass + " passed, " + fail + " failed");
        System.exit(fail > 0 ? 1 : 0);
    }
}

File: test/conformance/cpp/rows/W7.cpp
// CELL cpp/W7 — "arg and meta are two lanes" (docs/roadmap.sexp, row
// union-guards, audit schema#898 / matrix schema#876).
//
// THE LAW (docs/FIXED-FORM-ALGORITHM.md:240): "`arg` and `meta` ARE TWO LANES
// BECAUSE THEY ARE TWO FACTS, and must never share one. (fix 12) A `string(N)`
// under a union arm needs both, and one lane gives whichever was stamped
// last: a byte string under arm `2` read as WIDE, or an entry that runs only
// when the tag equals the flavour — under the wrong arm. Neither is visible to
// this build's own records; only to a peer's." The plan entry spells the two
// facts apart: `arg` IS THE GUARD'S ORDINAL and nothing else; `meta` IS THE
// OP'S OWN ARGUMENT (a text flavour: 1 utf8, 2 wide, 3 bytes) — the same
// paragraph's struct, docs/FIXED-FORM-ALGORITHM.md §4.1.
//
// WHY THIS FIXTURE (tables/backend/Backend.schema, the Envelope root): a
// union of three fixed tables whose arms carry a `bytes(32)` (LoginRequest.
// session_token) and a `string(32)` (StorePurchase.sku) as their text. The
// ordinal and the flavour DISAGREE at both text arms, which is the exact shape
// one lane cannot hold (the FM1/FU1 lesson, stated at test/tables/FM1.schema
// and test/tables/FU1.schema):
//
//   session_token  arm 1 (Login),    flavour 3 (bytes)  — arg 1, meta 3
//   sku            arm 3 (Purchase), flavour 1 (utf8)   — arg 3, meta 1
//
// Stamp one lane with "whichever was last" and either the guard becomes the
// flavour (session_token runs under arm 3 — under the wrong arm, and the login
// record reads empty) or the flavour becomes the guard's ordinal (session_token
// reads as utf8 and terminates the byte string, sku reads as bytes and leaves
// the C string unterminated). Both are visible on THIS build's own records the
// moment the baked identity plan is read back and a record is round-tripped
// through the production reader — the law says "only to a peer's" of the
// BUG, not of the assertion.
//
// WHAT THIS TEST DOES:
//   1. Holds EnvelopeFixedPlan's two kTableFixedText entries to the law: the
//      login arm's entry carries arg=Login and meta=kTableFixedTextBytes, the
//      purchase arm's carries arg=Purchase and meta=kTableFixedTextUtf8 — the
//      two facts disagree at both entries, and arg (u64) and meta (u8) are two
//      fields, never one. This is the generated-code constant the law governs.
//   2. Round-trips one record per text arm through the production writer and
//      the production reader (EnvelopeFixedSave / EnvelopeFixedLoad), with the
//      text's slack STAINED 0xFF after the save (a peer that leaves garbage
//      there is a peer this reader reads correctly — W4's law, SPEC-TABLES
//      §3.4), so the flavour's terminator write is observable:
//        - login / session_token (meta=bytes): NO terminator is written, so
//          the stained byte at the used length stays 0xFF. If meta were utf8
//          the terminator would land there and write a 0 — which is also what
//          the PURCHASE arm's entry would do to this byte if its guard were
//          wrong and it ran under the login tag (under the wrong arm).
//        - purchase / sku (meta=utf8): the terminator IS written at the used
//          length, so the stained byte there reads back 0. If meta were bytes
//          no terminator would be written and it would stay 0xFF.
//
// THE CONTROL (STEP 4, performed against build/tables-generated/, never this
// file): flip the ONE generated-code constant the law governs — session_token's
// plan-entry meta, 3 -> 1, the one-lane failure "a byte string read as utf8" —
// and this binary goes RED on both the plan assertion and the round-trip byte.
// Restore the constant: GREEN. (Breaking arg 1 -> 3 reddens the arg assertion
// and leaves the login record's session_token at its prefill: the entry runs
// under the wrong arm.)
//
// The wire is built by the generated writer (the conformance corpus has no
// fixed-form rows — W4's note holds here too) and the text payload's slack is
// stained at the fixed-form body's text run: body + src + 4, size 32, of which
// `length` bytes are the used units and 32 - length are the slack this read
// copies but must not act on (the flavour's terminator is the one store that
// lands there, and only when meta != bytes). src 9 / payload 13 / size 32 is
// EnvelopeFixedPlan's own text row for both arms (they share the wire offsets
// and differ in guard ordinal and flavour — that is the point).
#include <cstddef>
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <vector>

#include "BackendTable.h"

static int failures = 0;

static void check( bool ok, const char * what )
{
    std::printf( "%s %s\n", ok ? "PASS" : "FAIL", what );
    if ( !ok ) failures++;
}

using backenddemo::Envelope;
using backenddemo::EnvelopeFixedLoad;
using backenddemo::EnvelopeFixedMeasure;
using backenddemo::EnvelopeFixedPlan;
using backenddemo::EnvelopeFixedPlanCount;
using backenddemo::EnvelopeFixedSave;
using backenddemo::EnvelopeFixedLayoutBytes;
using backenddemo::LoginRequest;
using backenddemo::PayloadType;
using backenddemo::StorePurchase;
using backenddemo::TableFixedEntry;
using backenddemo::TableReport;
using backenddemo::kTableFixedHeaderBytes;
using backenddemo::kTableFixedText;
using backenddemo::kTableFixedTextBytes;
using backenddemo::kTableFixedTextUtf8;

// one record's body, after the fixed-form framing: header + 4 + layout + hash
static const uint8_t * body_of( const std::vector<uint8_t> & wire )
{
    return wire.data() + kTableFixedHeaderBytes + 4 + EnvelopeFixedLayoutBytes + 8;
}

// read one record off `wire` into `out`; returns the load's verdict
static int64_t read_one( const std::vector<uint8_t> & wire, Envelope & out, TableReport & r )
{
    backenddemo::EnvelopeReset( out );
    std::vector<TableFixedEntry> plan( 8192 );
    return EnvelopeFixedLoad( &out, 1, wire.data(), (int64_t) wire.size(),
                              plan.data(), (int32_t) plan.size(), NULL, &r );
}

// one Envelope record on the wire, with the text payload's slack stained 0xFF
// so a terminator write (meta != bytes) is visible against it. `login` selects
// the arm: login/session_token (bytes flavour) or purchase/sku (utf8 flavour).
static std::vector<uint8_t> build_wire( bool login )
{
    Envelope value;
    backenddemo::EnvelopeReset( value );
    if ( login )
    {
        value.payload.type = PayloadType::Login;
        value.payload.login = LoginRequest();
        value.payload.login.player_id = 0x1111111111111111ull;
        value.payload.login.session_token_length = 4;
        for ( int i = 0; i < 32; ++i ) { value.payload.login.session_token[i] = 0xFF; }
        value.payload.login.client_build = 0x22222222u;
    }
    else
    {
        value.payload.type = PayloadType::Purchase;
        value.payload.purchase = StorePurchase();
        value.payload.purchase.player_id = 0x3333333333333333ull;
        value.payload.purchase.sku_length = 4;
        for ( int i = 0; i < 32; ++i ) { value.payload.purchase.sku[i] = (char) 0xFF; }
        value.payload.purchase.sku[0] = 's';
        value.payload.purchase.sku[1] = 'k';
        value.payload.purchase.sku[2] = 'u';
        value.payload.purchase.sku[3] = '!';
        value.payload.purchase.quantity = 5;
        value.payload.purchase.price_minor = 0x44444444u;
    }
    std::vector<uint8_t> wire( (size_t) EnvelopeFixedMeasure( 1 ) );
    if ( EnvelopeFixedSave( &value, 1, wire.data(), (int64_t) wire.size() ) != (int64_t) wire.size() )
    {
        std::printf( "FAIL W7: the record does not save\n" );
        failures++;
        return wire;
    }
    // STAIN THE TEXT PAYLOAD'S SLACK (bytes 4..31 of the 32-byte run at
    // body + 13): 0xFF is a byte this reader must copy and never act on. The
    // one store that may land there is the flavour's terminator, and only
    // when meta != kTableFixedTextBytes. A peer leaving this garbage is a
    // peer this reader reads correctly (SPEC-TABLES §3.4, W4's row).
    std::memset( (uint8_t *) body_of( wire ) + 13 + 4, 0xFF, 32 - 4 );
    return wire;
}

int main()
{
    // ---- 1. THE PLAN CONSTANTS: arg and meta are two lanes ------------------
    //
    // TableFixedEntry lays arg (u64, the guard's ordinal) and meta (u8, the
    // text flavour) in two separate fields — the two lanes the law names. A
    // ten-value aggregate that shared one would have to choose between "which
    // arm am I guarded by" and "which flavour of text am I" for a string(N)
    // under an arm, and the FM1/FU1 fixtures record what that choice cost.
    check( offsetof( TableFixedEntry, arg ) != offsetof( TableFixedEntry, meta ),
           "W7: arg and meta are two lanes (two fields, never one)" );
    check( sizeof( ( (TableFixedEntry *) nullptr )->arg ) == 8 && sizeof( ( (TableFixedEntry *) nullptr )->meta ) == 1,
           "W7: arg is full width (bill §12.7) and meta is the op's own byte" );

    int text_entries = 0;
    bool saw_login_text = false;
    bool saw_purchase_text = false;
    for ( int32_t i = 0; i < EnvelopeFixedPlanCount; ++i )
    {
        const TableFixedEntry & p = EnvelopeFixedPlan[i];
        if ( p.op != kTableFixedText ) { continue; }
        text_entries++;
        check( p.guard == 0, "W7: a text entry under an arm guards on the payload tag (offset 0)" );
        if ( p.arg == (uint64_t) PayloadType::Login )
        {
            saw_login_text = true;
            // arm 1 (Login), flavour 3 (bytes): the ordinal and the flavour
            // DISAGREE — one lane cannot hold both facts.
            check( p.meta == (uint8_t) kTableFixedTextBytes,
                   "W7: session_token carries the Login ordinal in arg AND the bytes flavour in meta (two lanes)" );
            check( p.arg != (uint64_t) p.meta,
                   "W7: session_token's guard ordinal and flavour disagree (one lane could not hold both)" );
        }
        else if ( p.arg == (uint64_t) PayloadType::Purchase )
        {
            saw_purchase_text = true;
            // arm 3 (Purchase), flavour 1 (utf8): the ordinal and the flavour
            // DISAGREE at this arm too.
            check( p.meta == (uint8_t) kTableFixedTextUtf8,
                   "W7: sku carries the Purchase ordinal in arg AND the utf8 flavour in meta (two lanes)" );
            check( p.arg != (uint64_t) p.meta,
                   "W7: sku's guard ordinal and flavour disagree (one lane could not hold both)" );
        }
    }
    check( text_entries == 2, "W7: EnvelopeFixedPlan has exactly two text entries (one per text arm)" );
    check( saw_login_text && saw_purchase_text, "W7: both text arms' entries are on the plan" );

    // ---- 2. THE ROUND TRIP: each lane does its own job ----------------------
    //
    // Login / session_token — meta=bytes: no terminator, and the stained byte
    // at the used length must survive. The arg lane is what runs this entry
    // under tag=Login only (and keeps the purchase arm's utf8 entry off it).
    {
        const std::vector<uint8_t> wire = build_wire( true );
        Envelope out;
        TableReport r;
        const int64_t n = read_one( wire, out, r );
        check( n == 1 && !r.refused && !r.malformed && r.clamped == 0,
               "W7: the login record reads clean" );
        check( out.payload.type == PayloadType::Login, "W7: the login tag lands (arg's guard matched)" );
        check( out.payload.login.player_id == 0x1111111111111111ull, "W7: login.player_id lands" );
        check( out.payload.login.session_token_length == 4, "W7: session_token_length lands under the Login arm (arg=1)" );
        check( out.payload.login.session_token[0] == 0xFF && out.payload.login.session_token[3] == 0xFF,
               "W7: session_token's used units land" );
        // THE FLAVOUR'S BITE: meta=bytes writes NO terminator, so the stained
        // slack byte at the used length is still 0xFF. meta=utf8 would have
        // written 0 there — and so would the purchase arm's entry, if its
        // guard were wrong and it ran under this tag (under the wrong arm).
        check( out.payload.login.session_token[4] == 0xFF,
               "W7: bytes flavour writes no terminator (meta=bytes, arg and meta are two lanes)" );
        check( out.payload.login.client_build == 0x22222222u, "W7: the login arm's neighbour after the text lands" );
    }

    // Purchase / sku — meta=utf8: the terminator IS written at the used
    // length, so the stained byte there reads back 0. meta=bytes would have
    // left it 0xFF and the C string would be unterminated at its used length.
    {
        const std::vector<uint8_t> wire = build_wire( false );
        Envelope out;
        TableReport r;
        const int64_t n = read_one( wire, out, r );
        check( n == 1 && !r.refused && !r.malformed && r.clamped == 0,
               "W7: the purchase record reads clean" );
        check( out.payload.type == PayloadType::Purchase, "W7: the purchase tag lands (arg's guard matched)" );
        check( out.payload.purchase.player_id == 0x3333333333333333ull, "W7: purchase.player_id lands" );
        check( out.payload.purchase.sku_length == 4, "W7: sku_length lands under the Purchase arm (arg=3)" );
        check( out.payload.purchase.sku[0] == 's' && out.payload.purchase.sku[1] == 'k' &&
               out.payload.purchase.sku[2] == 'u' && out.payload.purchase.sku[3] == '!',
               "W7: sku's used units land" );
        // THE FLAVOUR'S BITE: meta=utf8 terminates at the used length.
        check( out.payload.purchase.sku[4] == 0,
               "W7: utf8 flavour writes its terminator (meta=utf8, arg and meta are two lanes)" );
        check( out.payload.purchase.quantity == 5, "W7: the purchase arm's neighbour after the text lands" );
    }

    return failures == 0 ? 0 : 1;
}
